readme.md
# MMM-LOLESPORTS-SCHEDULES
An Esports schedule module for a provided League of Legends tournament id.
Module for <a href="https://github.com/MichMich/MagicMirror">MagicMirror</a>.

## Previews with Configuration Samples
#### Default
<img src="https://github.com/xadamxk/MMM-LOLESPORTS-SCHEDULES/blob/main/screenshots/default_live.png?raw=true" title="Default Configuration Preview"  />

<details> 
  <summary>Hide Team Label:</summary>
  <img src="https://github.com/xadamxk/MMM-LOLESPORTS-SCHEDULES/blob/main/screenshots/hideFullName.png?raw=true" title="Preview Hide Team Label"  />
	<pre><code>
config: {
	showTeamLabel: false	
}
	</code></pre>
</details>

<details> 
  <summary>Minified Layout:</summary>
  <img src="https://github.com/xadamxk/MMM-LOLESPORTS-SCHEDULES/blob/main/screenshots/hideFullName_hideRegionAndBestOf.png?raw=true" title="Preview Minified Layout"  />
	<pre><code>
config: {
	showTeamLabel: false,
	showRegionAndBestOf: false,
}
	</code></pre>
</details>

<details> 
  <summary>Full Layout:</summary>
  <img src="https://github.com/xadamxk/MMM-LOLESPORTS-SCHEDULES/blob/main/screenshots/hidePeriodLabel_use24hour_useFullTeamName_showTournamentBlock.png?raw=true" title="Preview Full Layout"  />
	<pre><code>
config: {
	showPeriodLabel: false,
	use24HourTime: true,
	useTeamFullName: true,
	showTournamentBlock: true
}
	</code></pre>
</details>

## Installation
In your terminal, go to your MagicMirror's Module folder:
````
cd ~/MagicMirror/modules
````

Clone this repository:
````
git clone https://github.com:xadamxk/MMM-LOLESPORTS-SCHEDULES.git
````

## Configuration
Add `MMM-LOLESPORTS-SCHEDULE` module to the `modules` array in the `config/config.js` file:
````javascript
modules: [
	{
		module: "MMM-LOLESPORTS-SCHEDULES",
		position: "middle_center",
		config: {}
	}
]
````
### Configuration Options

| **Option** | **Type** | **Default** | **Description** |
| --- | --- | --- | --- |
| `updateInterval` | integer | 60 | Number of minutes to poll api for updates. |
| `apiKey` | string | '0TvQnueqKa5mxJntVWt0w4LpLfEkrV1Ta8rQBb9Z' | Api key used to query esports API - all users' api key is the default key. |
| `basePath` | string | 'https://esports-api.lolesports.com/persisted/gw' | Base bath used to query the esports api. |
| `leagueId` | array | ["98767991299243165"] | Array of league ids to get esport standings. Currently the API only supports a single league id. If no id is provided, then all region games will be provided. Refer to league table below for ids of other leagues. Defaults to NA LCS.|
| `hl` | string | 'en-US' | Host language/ locale to use when requesting esports data. |
| `numberOfFutureGames` | integer | `5` | Number of future games to display. |
| `use24HourTime` | boolean | false | Use 24 hour format (hh:mm) rather than 12 hour format (hh PERIOD) |
| `useTeamFullName` | boolean | false | Show teams' full name rather than teams' code. |
| `showPeriodLabel` | boolean | true | Show period (AM/PM) after time. |
| `showTeamLabel` | boolean | true | Show team label (name/code). Use `false` for slim component. |
| `showRegionAndBestOf` | boolean | true | Show region and best of format for games. |
| `showTournamentBlock` | boolean | false | Show tournament block (ie. Playoffs - Round 1) after start date.|

### League Ids
League Ids are obtained by navigating to `https://lolesports.com/schedule`, opening dev tools, selecting a league filter, and finding the league id in the network panel.

| **League** | **League Id** |
| --- | --- |
| LCS | `98767991299243165` |
| LCS Academy | `99332500638116286` |
| LEC | `98767991302996019` |
| LCK | `98767991310872058` |
| LPL | `98767991314006698` |
| TCL | `98767991343597634` |
| CBLOL | `98767991332355509` |
| LLA | `101382741235120470` |
| LCO | `105709090213554609` |
| LJL | `98767991349978712` |
| LCL | `98767991355908944` |



translations/CHANGELOG.md
## [1.0.0] - 2021-xx-xx
- First public release

node_helper.js
/* Magic Mirror
 * Node Helper: MMM-LOLESPORTS-SCHEDULES
 *
 * By xadamxk
 * MIT Licensed.
 */

var NodeHelper = require("node_helper");
var LolesportsApi = require("esm")(module)("lolesports-js-sdk");
var defaultClient = LolesportsApi.ApiClient.instance;
var api = new LolesportsApi.EventsApi();

module.exports = NodeHelper.create({
  // Override socketNotificationReceived method.
  socketNotificationReceived: function (notification, payload) {
    let self = this;
    if (notification === "MMM-LOLESPORTS-SCHEDULES-GET-SCHEDULE") {
      defaultClient.basePath = payload["basePath"];
      defaultClient.authentications["apiKeyAuth"]["apiKey"] = payload["apiKey"];

      api.getSchedule(
        LolesportsApi.Locale.enUS,
        { leagueId: payload["leagueId"] },
        function (error, data, response) {
          if (error) {
            console.error(error);
          } else {
            // console.log("API called successfully. Returned data: " + JSON.stringify(data));
            self.sendNotificationTest(data);
          }
        }
      );
    }
  },
  // Example function send notification test
  sendNotificationTest: function (payload) {
    this.sendSocketNotification("MMM-LOLESPORTS-SCHEDULES-SCHEDULE", payload);
  }
});


translations/MMM-LOLESPORTS-STANDINGS.js
Module.register("MMM-LOLESPORTS-SCHEDULES", {
  // Default module config
  defaults: {
    updateInterval: 60, // minutes
    // lang: config.language,
    apiKey: "0TvQnueqKa5mxJntVWt0w4LpLfEkrV1Ta8rQBb9Z",
    basePath: "https://esports-api.lolesports.com/persisted/gw",
    tournamentIds: ["105658534671026792"], // NA LCS
    hl: "en-US",
    useTeamFullName: true, // Show team's full name rather than team code
    showTeamIcons: true, // Show team's icon
    showStageName: true // Show the stage name (ie. Regular Season, Playoffs, etc)
  },

  // Module properties.
  standings: [],
  stageName: "",

  // Start the module.
  start: function () {
    // Get initial API data
    this.getData();

    // Schedule update poll
    var self = this;
    setInterval(function () {
      self.getData();
    }, self.config.updateInterval * 60 * 1000); //convert to milliseconds
  },
  getTranslations() {
    return {
      en: "translations/en.json"
    };
  },
  getStyles: function () {
    return ["MMM-LOLESPORTS-SCHEDULES.css"];
  },
  getTemplate() {
    return `templates/schedules.njk`;
  },
  getTemplateData() {
    return {
      standings: this.standings,
      stageName: this.stageName,
      config: this.config
    };
  },
  // Fetch schedule for provided tournament ids
  getData: function () {
    this.sendSocketNotification("MMM-LOLESPORTS-SCHEDULES-GET-SCHEDULE", {
      apiKey: this.config.apiKey,
      basePath: this.config.basePath,
      tournamentIds: this.config.tournamentIds,
      hl: this.config.hl
    });
  },

  // Schedule data is coming back
  socketNotificationReceived: function (notification, payload) {
    if (notification === "MMM-LOLESPORTS-SCHEDULES-SCHEDULE") {
      this.getStandingData(payload);
    }
  },
  // Condense standing data and render it
  getStandingData: function (data) {
    let stageName = "";
    if (!data || !data.hasOwnProperty("data")) {
      return []; // Wrong tournament id most likely
    }
    data["data"]["standings"].forEach((standing) => {
      if (!standing["stages"]) {
        return;
      }
      standing["stages"].forEach((stage) => {
        stageName = stage["name"];
        stage["sections"].forEach((section) => {
          if (!section["rankings"] || !section["rankings"].length) {
            return;
          }
          this.standings = section["rankings"];
          this.stageName = stageName;
          this.updateDom(500);
        });
      });
    });
  }
});


MMM-LOLESPORTS-SCHEDULES.js
Module.register("MMM-LOLESPORTS-SCHEDULES", {
  // Default module config
  defaults: {
    updateInterval: 60, // minutes
    // lang: config.language,
    apiKey: "0TvQnueqKa5mxJntVWt0w4LpLfEkrV1Ta8rQBb9Z",
    basePath: "https://esports-api.lolesports.com/persisted/gw",
    hl: "en-US",
    // Custom
    leagueId: ["98767991299243165"], // NA LCS
    numberOfFutureGames: 5,
    use24HourTime: false,
    useTeamFullName: false,
    showPeriodLabel: true,
    showTeamLabel: true,
    showRegionAndBestOf: true,
    showTournamentBlock: false
  },

  // Module properties.
  groupedSchedules: [],

  // Start the module.
  start: function () {
    // Get initial API data
    this.getData();

    // Schedule update poll
    var self = this;
    setInterval(function () {
      self.getData();
    }, self.config.updateInterval * 60 * 1000); //convert to milliseconds
  },
  getTranslations() {
    return {
      en: "translations/en.json"
    };
  },
  getStyles: function () {
    return ["MMM-LOLESPORTS-SCHEDULES.css"];
  },
  getTemplate() {
    return `templates/schedules.njk`;
  },
  getTemplateData() {
    return {
      groupedSchedules: this.groupedSchedules,
      config: this.config
    };
  },
  // Fetch schedule for provided league ids
  getData: function () {
    this.sendSocketNotification("MMM-LOLESPORTS-SCHEDULES-GET-SCHEDULE", {
      apiKey: this.config.apiKey,
      basePath: this.config.basePath,
      leagueId: this.config.leagueId,
      hl: this.config.hl
    });
  },

  // Schedule data is coming back
  socketNotificationReceived: function (notification, payload) {
    if (notification === "MMM-LOLESPORTS-SCHEDULES-SCHEDULE") {
      this.getSchedulesData(payload);
    }
  },
  // Condense schedule data and render it
  getSchedulesData: function (data) {
    if (!data || !data.hasOwnProperty("data")) {
      return []; // Wrong league id most likely
    }
    if (!data["data"]["schedule"] || !data["data"]["schedule"]["events"]) {
      return []; // No events for league id
    }
    const events = data["data"]["schedule"]["events"];

    // Date helpers
    const getMidnight = (day) => {
      const date = new Date(day);
      date.setMilliseconds(999);
      date.setSeconds(59);
      date.setMinutes(59);
      date.setHours(23);
      return date;
    };

    const oneDay = 1000 * 60 * 60 * 24;

    const isToday = (date) => {
      const today = new Date();
      return (
        // eslint-disable-next-line eqeqeq
        date.getDate() == today.getDate() &&
        // eslint-disable-next-line eqeqeq
        date.getMonth() == today.getMonth() &&
        // eslint-disable-next-line eqeqeq
        date.getFullYear() == today.getFullYear()
      );
    };

    const isTomorrow = (date) => {
      const midnightTonight = getMidnight(new Date());
      const midnightTomorrow = new Date(midnightTonight.getTime() + oneDay);
      return date > midnightTonight && date < midnightTomorrow;
    };

    const futureEvents = events
      .filter((event) => {
        return (
          event["startTime"] > new Date().toISOString() ||
          event["state"] === "inProgress"
        );
      })
      .slice(0, this.config.numberOfFutureGames)
      // Add custom fields for display
      .map((event) => {
        // Is Live
        if (event["state"] === "inProgress") {
          event["isLive"] = true;
        } else {
          event["isLive"] = false;
        }
        // Start time fields
        const startTimeDate = new Date(event["startTime"]);
        let hours = startTimeDate.getHours();
        if (hours >= 12) {
          event["startPeriod"] = "PM";
        } else {
          event["startPeriod"] = "AM";
        }
        if (!this.config.use24HourTime) {
          hours = hours % 12;
        }
        event["startHour"] = hours;
        // Start date fields
        if (isToday(startTimeDate)) {
          event["startDayOfWeek"] = this.translate("TODAY");
        } else if (isTomorrow(startTimeDate)) {
          event["startDayOfWeek"] = this.translate("TOMORROW");
        } else {
          switch (startTimeDate.getDay()) {
            case 0:
              event["startDayOfWeek"] = this.translate("SUNDAY");
              break;
            case 1:
              event["startDayOfWeek"] = this.translate("MONDAY");
              break;
            case 2:
              event["startDayOfWeek"] = this.translate("TUESDAY");
              break;
            case 3:
              event["startDayOfWeek"] = this.translate("WEDNESDAY");
              break;
            case 4:
              event["startDayOfWeek"] = this.translate("THURSDAY");
              break;
            case 5:
              event["startDayOfWeek"] = this.translate("FRIDAY");
              break;
            case 6:
              event["startDayOfWeek"] = this.translate("SATURDAY");
              break;
            default:
              event["startDayOfWeek"] = this.translate("SUNDAY");
          }
        }
        // Start month fields
        switch (startTimeDate.getMonth()) {
          case 0:
            event["startMonth"] = this.translate("JANUARY");
            break;
          case 1:
            event["startMonth"] = this.translate("FEBRUARY");
            break;
          case 2:
            event["startMonth"] = this.translate("MARCH");
            break;
          case 3:
            event["startMonth"] = this.translate("APRIL");
            break;
          case 4:
            event["startMonth"] = this.translate("MAY");
            break;
          case 5:
            event["startMonth"] = this.translate("JUNE");
            break;
          case 6:
            event["startMonth"] = this.translate("JULY");
            break;
          case 7:
            event["startMonth"] = this.translate("AUGUST");
            break;
          case 8:
            event["startMonth"] = this.translate("SEPTEMBER");
            break;
          case 9:
            event["startMonth"] = this.translate("OCTOBER");
            break;
          case 10:
            event["startMonth"] = this.translate("NOVEMBER");
            break;
          case 11:
            event["startMonth"] = this.translate("DECEMBER");
            break;
          default:
            event["startMonth"] = this.translate("JANUARY");
            break;
        }

        event["startDayOfMonth"] = startTimeDate.getDate();

        return event;
      });

    // Credit: https://stackoverflow.com/a/37844673/2694643
    var groupByDay = function (obj) {
      var objPeriod = {};
      var oneDay = 24 * 60 * 60 * 1000; // hours * minutes * seconds * milliseconds
      for (var i = 0; i < obj.length; i++) {
        var d = new Date(obj[i]["startTime"]); // unique to schedule data
        d = Math.floor(d.getTime() / oneDay);
        // define object key
        objPeriod[d] = objPeriod[d] || [];
        objPeriod[d].push(obj[i]);
      }
      return objPeriod;
    };
    // Games grouped by day
    const groupedSchedulesMap = groupByDay(futureEvents);
    const groupedSchedules = Object.values(groupedSchedulesMap).map(
      (values) => values
    );
    this.groupedSchedules = groupedSchedules;
    this.updateDom(500);
  }
});
